pkg/distributor/receiver/shim.go

```diff
@@ -0,0 +1,318 @@
+		Receivers: receiverFactories,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+	params := component.ReceiverCreateSettings{TelemetrySettings: component.TelemetrySettings{
+		Logger:         zap.NewNop(),
+		TracerProvider: trace.NewNoopTracerProvider(),
+		MeterProvider:  metric.NewNoopMeterProvider(),
+	}}
+
+	for componentID, cfg := range cfgs.Receivers {
+		factoryBase := receiverFactories[componentID.Type()]
+		if factoryBase == nil {
+			return nil, fmt.Errorf("receiver factory not found for type: %s", componentID.Type())
+		}
+
+		receiver, err := factoryBase.CreateLogsReceiver(ctx, params, cfg, middleware.Wrap(shim))
+		if err != nil {
+			return nil, err
+		}
+
+		shim.receivers = append(shim.receivers, receiver)
+	}
+
+	shim.Service = services.NewIdleService(shim.starting, shim.stopping)
+
+	return shim, nil
+}
+func (r *receiversShim) starting(ctx context.Context) error {
+	for _, receiver := range r.receivers {
+		r.logger.Log("msg", "receiver start")
+		err := receiver.Start(ctx, r)
+		if err != nil {
+			return fmt.Errorf("error starting receiver %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Called after distributor is asked to stop via StopAsync.
+func (r *receiversShim) stopping(_ error) error {
+	// when shutdown is called on the receiver it immediately shuts down its connection
+	// which drops requests on the floor. at this point in the shutdown process
+	// the readiness handler is already down . so we are not receiving any more requests.
+	// sleep for 30 seconds to here to all pending requests to finish.
+	time.Sleep(r.drainTimeout)
+
+	ctx, cancelFn := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancelFn()
+
+	errs := make([]error, 0)
+
+	for _, receiver := range r.receivers {
+		err := receiver.Shutdown(ctx)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 0 {
+		return multierr.Combine(errs...)
+	}
+
+	return nil
+}
+
+// ConsumeLogs implements consumer.Logs
+func (r *receiversShim) ConsumeLogs(ctx context.Context, ld pdata.Logs) error {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "distributor.ConsumeLogs")
+	defer span.Finish()
+	req, err := parseLog(ld, r.format)
+	if err != nil {
+		r.logger.Log("msg", "pusher failed to parse log data", "err", err)
+	}
+
+	start := time.Now()
+	_, err = r.pusher.Push(ctx, req)
+	metricPushDuration.Observe(time.Since(start).Seconds())
+	if err != nil {
+		r.logger.Log("msg", "pusher failed to consume log data", "err", err)
+	}
+
+	return err
+}
+
+func parseLog(ld pdata.Logs, format string) (*logproto.PushRequest, error) {
+	streams := make(map[string]*logproto.Stream)
+	rss := ld.ResourceLogs()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		ill := rs.InstrumentationLibraryLogs()
+		for i := 0; i < ill.Len(); i++ {
+			logs := ill.At(i).LogRecords()
+			for k := 0; k < logs.Len(); k++ {
+				pLog := logs.At(k)
+				labels := parseLabel(rs.Resource().Attributes(), pLog.Attributes())
```

Comment:
I'll get there but wanted to mention this before I forget: we have an accept-list approach on the otelcol side of things: users have to specify which record, resource, or log attributes they want to leave as labels. Everything else ends up part of the body.

Example:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6619